### PR TITLE
Podcast Player - make feed fetch call cancellable and cancel on Block removal.

### LINKED ISSUE #15214: Podcast Player: error if Block unmounts whilst fetching feed.
Currently the Podcast Player Block uses `apiFetch` to trigger a Ajax request for the feed data. Once it [the fetch promise] resolves then state and attributes updates are triggered.

If, whilst the network request is in progress, the Block is "unmounted" then the console will show an error along the lines of

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.

This is because the `.then` handler on the fetch still triggers even though the Block no longer exists.

##### Steps to reproduce the issue

* Insert Podcast Block.
* In Network Tools switch to "Slow 3G mode".
* Add a valid feed URL and submit.
* Immediately (before the fetch request resolves) delete the Block.
* See error in console. 

#### What I expected

If the Block is removed in the middle of the fetching of the feed then there should be no errors in the console and everything should be handled gracefully. 

#### What happened instead

There are errors in the console.

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.

#### Solution

Please see the `LinkControl` component from Gutenberg. In this, I encountered a similar problem which I solved by [wrapping the Promise to allow it to be cancelable](https://github.com/WordPress/gutenberg/blob/ddd3fffdd327d2f9c5202481345fb7427d4a822b/packages/block-editor/src/components/link-control/index.js#L62).

The cancellables then need to be [stored in refs](https://github.com/WordPress/gutenberg/blob/ddd3fffdd327d2f9c5202481345fb7427d4a822b/packages/block-editor/src/components/link-control/index.js#L167) to ensure they can be handled.

Then [when the block is "unmounted" all cancellables should be `.cancel()`](https://github.com/WordPress/gutenberg/blob/ddd3fffdd327d2f9c5202481345fb7427d4a822b/packages/block-editor/src/components/link-control/index.js#L365).

I would also like to experiment to see if we can extract this to a hook and then see if @aduth would be interested in having it in Gutenberg because he and I discussed that this might be an issue for 3rd party Blocks.

### PR DESCRIPTION
**Pull request by getdave**
Currently the Podcast Player Block uses apiFetch to trigger a Ajax request for the feed data. Once it [the fetch promise] resolves then state and attributes updates are triggered.

If, whilst the network request is in progress, the Block is "unmounted" then the console will show an error along the lines of

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.

This is because the .then handler on the fetch still triggers even though the Block no longer exists. As the handler updates state and attributes the console triggers the above error regarding state updates on unmounted components.

To fix this we make the `apiFetch` call cancell-able (via a wrapper util) and then ensure we cancel and do not update state if the Block has been unmounted.

Fixes https://github.com/Automattic/jetpack/issues/15214

## Changes proposed in this Pull Request:
* Wrap `apiFetch` to make it possible to cancel it.
* Add ref to track cancellable promise instance.
* Add `useEffect` to `.cancel()` the fetch promise if the Block unmounts.
* Add conditional to the `.catch()` handler to check for `.isCancelled` and terminate handler execution early to avoid state updates.

## Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Fixes existing bug in Podcast Player Block

## Testing instructions:

Annoyingly there seems to be a bug in Gutenberg with the Popover component. Therefore when "removing" the BLock below you will need to click to place the cursor in the editor immediate following the Block and then hit the `BACKSPACE` key to remove the block. Removing using the dropdown many on the Block itself will trigger a warning which is **unrelated to this Block**.

#### On Feature Branch

* Insert Podcast Block.
* In Network Tools switch to "Slow 3G mode".
* Add a valid feed URL and submit.
* Immediately (before the fetch request resolves) delete the Block (see notes above).
* See error in console relating to state updates on unmounted component.

#### On This Branch

* Insert Podcast Block.
* In Network Tools switch to "Slow 3G mode".
* Add a valid feed URL and submit.
* Immediately (before the fetch request resolves) delete the Block (see notes above).
* See no errors.

If you enable `debug` mode you will also see a notice informing you that:

> Block was unmounted during fetch

 

#### Proposed changelog entry for your changes:
* Fix React error regarding state updates on unmounted component if Block is removed during the fetching of the Podcast feed.

### CODE REVIEW

File: extensions/blocks/podcast-player/edit.js [hunk 1]
```diff
@@ -81,21 +82,30 @@ const PodcastPlayerEdit = ( {
...
-				path: '/wpcom/v2/podcast-player?url=' + encodedURL,
-			} ).then(
+			cancellableFetch.current = makeCancellable(
+				apiFetch( {
+					path: '/wpcom/v2/podcast-player?url=' + encodedURL,
+				} )
+			);
+
+			cancellableFetch.current.promise.then(
 				data => {
```
Comment (retrofox) [inline extensions/blocks/podcast-player/edit.js:98]: `isCancel` seems to be present in the `data` object too. Does it make sense to check if the request has been canceled here in the same way when an error happens?

File: extensions/blocks/podcast-player/edit.js [hunk 2]
```diff
@@ -106,6 +116,14 @@ const PodcastPlayerEdit = ( {
...
 	);
 
+	useEffect( () => {
+		return () => {
+			if ( cancellableFetch && cancellableFetch.current && cancellableFetch.current.cancel ) {
+				cancellableFetch.current.cancel();
+			}
+		};
+	}, [] );
+
```
Comment (retrofox) [inline extensions/blocks/podcast-player/edit.js:126]: Simple suggestion to make it clearer. Feel free to ignore it :-)

```suggestion
	useEffect( () => (
		() => get( cancellableFetch, [ 'current', 'cancel' ] )
			? cancellableFetch.current.cancel()
			: noop()
	), [] );
```

Don't forget to add the dependencies if you agree with these changes.

**Review by retrofox — APPROVED** [review]
I left only one comment about checking the `isCancel` state. Maybe it doesn't make sense, though. 

Tested and it works as expected.

image

Nice improvement, Dave. 💯

We can wait for more feedback from other folks.

File: extensions/blocks/podcast-player/edit.js (hunk 2 above)
Comment (getdave) [reply extensions/blocks/podcast-player/edit.js]: OK I'm not 100% sure on this but I've ended up using Optional Chaining as an alternative here.

Supported as of https://github.com/Automattic/wp-calypso/blob/c3608b5c2b656cad1209f52b0e2229838c2592ea/packages/calypso-build/CHANGELOG.md#510

File: extensions/blocks/podcast-player/edit.js (hunk 1 above)
Comment (getdave) [reply extensions/blocks/podcast-player/edit.js:98]: Great spot. You are absolutely correct. I've fixed this also.

**Comment by getdave** [comment]
As this is merging into the feature branch we won't wait on Jetpack Crew.

**Pull request merged**